bc_obps/service/form_builder_service.py

```diff
@@ -0,0 +1,210 @@
+                        gas_type_id=t.gas_type.id,
+                        methodology_id=u.methodology.id,
+                        valid_from__lte=config,
+                        valid_to__gte=config,
+                    )
+                    .reporting_fields.all()
+                )
+            except:
+                raise Exception(
+                    f'No configuration found for activity_id {activity} & source_type_id {source_type} & gas_type_id {t.gas_type.id} & methodology_id {u.methodology.id} & report_date {report_date}'
+                )
+            methodology_object:Any = {"properties": {"methodology": {"enum": [u.methodology.name]}}}
+            # For each field related to a methodology, add that field to the methodology object & append to the oneOf branch
+            for f in methodology_fields:
+                property_field = str_to_camel_case(f.field_name)
+                methodology_object['properties'][property_field] = {"type": f.field_type, "title": f.field_name}
+                if f.field_units:
+                    methodology_object['properties'][f"{property_field}FieldUnits"] = {
+                        "type": "string",
+                        "default": f.field_units,
+                        "title": f"{f.field_name} Units",
+                    }
+            methodology_one_of['methodology']['oneOf'].append(methodology_object)
+        gas_type_one_of['gasType']['oneOf'][index]['properties']['methodology']['enum'] = methodology_enum
+        gas_type_one_of['gasType']['oneOf'][index]['dependencies'] = methodology_one_of
+
+    st_schema:Any = source_type_schema.json_schema
+    # Append valid gas types to schema as an enum on the gasType property. Uses the has_unit / has_fuel booleans to determine the depth of the emissions array.
+    if source_type_schema.has_unit and source_type_schema.has_fuel:
+        st_schema['properties']['units']['items']['properties']['fuels']['items']['properties']['emissions']['items'][
+            'properties'
+        ]['gasType']['enum'] = gas_type_enum
+        st_schema['properties']['units']['items']['properties']['fuels']['items']['properties']['emissions']['items'][
+            'dependencies'
+        ] = gas_type_one_of
+    elif not source_type_schema.has_unit and source_type_schema.has_fuel:
+        st_schema['properties']['fuels']['items']['properties']['emissions']['items']['properties']['gasType'][
+            'enum'
+        ] = gas_type_enum
+        st_schema['properties']['fuels']['items']['properties']['emissions']['items']['dependencies'] = gas_type_one_of
+    else:
+        st_schema['properties']['emissions']['items']['properties']['gasType']['enum'] = gas_type_enum
+        st_schema['properties']['emissions']['items']['dependencies'] = gas_type_one_of
+    # Add the source_type schema to the schema object being returned by this function
+    rjsf_schema['properties']['sourceTypes']['properties'][
```

Comment:
Instead of passing in the top-level schema and making this function dependent on the data that is passed in originally, I would return `st_schema` and let the calling method put the returned schema data in the right place.

We could make this function return a tuple so that the caller also gets the name of the source type